File: bff-api/internal/services/group_service.go
package services

import (
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/maxiking445/bff-api/internal/common"
	model "github.com/maxiking445/bff-api/internal/models"
)

func LoadGroupsWithMessageCounts(groupsPath string) ([]model.Group, error) {
	groups, err := loadGroupsFromCSV(groupsPath)
	if err != nil {
		return nil, err
	}

	files, err := filepath.Glob(common.GroupMessagePath)
	if err != nil {
		panic(err)
	}

	var msgs []GroupMessageRow
	for _, path := range files {
		fmt.Println("Analyse Groupe File: " + path)
		rows, err := loadGroupMessagesFromCSV(path)
		if err != nil {
			return nil, err
		}
		msgs = append(msgs, rows...)

	}

	groupByUID := countIdentitiesPerGroup(msgs)
	// Create GroupMemebr sub Object
	for i := range groups {
		g := &groups[i] // Pointer auf das Original
		m := groupByUID[g.GroupUID]
		for identity := range m {
			groupMember := model.GroupMember{
				MessageCount: int64(m[identity]),
				Identity:     LoadIdentity(identity),
			}
			g.GroupMember = append(g.GroupMember, groupMember)
		}
	}
	//TOTAL COUND CALC
	for i := range groups {
		g := &groups[i]
		g.MessageCount = 0
		for _, s := range g.GroupMember {
			g.MessageCount += int64(s.MessageCount)
		}
	}

	return groups, nil
}

// Count Messages
func countIdentitiesPerGroup(rows []GroupMessageRow) map[string]map[string]int {
	result := make(map[string]map[string]int)

	for _, r := range rows {
		if result[r.GroupUID] == nil {
			result[r.GroupUID] = make(map[string]int)
		}

		result[r.GroupUID][r.Identity]++
	}

	return result
}

func loadGroupsFromCSV(path string) ([]model.Group, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	reader := csv.NewReader(f)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	if len(records) <= 1 {
		return []model.Group{}, nil
	}

	var groups []model.Group
	for i, r := range records {
		if i == 0 {
			continue
		}

		if len(r) < 11 {
			continue
		}

		archived, _ := strconv.ParseBool(r[6])

		groupItem := model.Group{
			ID:        r[0],
			Creator:   r[1],
			GroupName: r[2],
			Archived:  archived,
			GroupUID:  r[9],
		}

		groups = append(groups, groupItem)
	}

	return groups, nil
}

type GroupMessageRow struct {
	GroupUID  string
	Identity  string
	Timestamp time.Time
}

func loadGroupMessagesFromCSV(path string) ([]GroupMessageRow, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	reader := csv.NewReader(f)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	if len(records) <= 1 {
		return []GroupMessageRow{}, nil
	}

	var rows []GroupMessageRow
	for i, r := range records {
		if i == 0 {
			continue
		}

		if len(r) < 12 {
			continue
		}
		if r[10] != "TEXT" {
			continue
		}

		base := filepath.Base(path)
		base = strings.TrimSuffix(base, common.CSV)
		groupUUID := strings.TrimPrefix(base, common.GroupMessage)

		createdAtMillis, err := strconv.ParseInt(r[7], 10, 64)
		if err != nil {
			fmt.Println("Error loading created_at:", r[7], "Error:", err)
			continue
		}
		timestamp := time.UnixMilli(createdAtMillis)

		identity := r[2]
		if identity == "" {
			identity = "YOU"
		}

		rows = append(rows, GroupMessageRow{
			GroupUID:  groupUUID, // uid = group_uid
			Identity:  identity,  // identity = user
			Timestamp: timestamp, //Timestamp
		})
	}

	return rows, nil
}

File: analyzer-frontend/src/service/GroupService.ts
// services/group-service.ts
import { getZipFile, listFiles } from "./FileService"; 
import Papa from "papaparse"; 
import { loadIdentityByUserID } from "./IdentityService";
import { ModelsGroup, ModelsGroupMember } from "@/models";

interface GroupMessageRow {
  groupUID: string;
  identity: string;
  timestamp: Date;
}

export async function loadGroupsWithMessageCounts(
  groupsFilename: string,
): Promise<ModelsGroup[]> {
  const groupsFile = getZipFile(groupsFilename);
  if (!groupsFile) {
    throw new Error(`Groups file "${groupsFilename}" not found`);
  }

  const groups = await loadGroupsFromCSV(groupsFile.content);

  const messageFiles = listFiles().filter(
    (f) => f.startsWith("group_message_") && f.endsWith(".csv"),
  );

  let allMessages: GroupMessageRow[] = [];

  // 3. Alle Message Files laden
  for (const filename of messageFiles) {
    console.log("Analyse Group File:", filename);
    const fileContent = getZipFile(filename);
    if (fileContent) {
      const messages = await loadGroupMessagesFromCSV(
        fileContent.content,
        filename,
      );
      allMessages.push(...messages);
    }
  }

  // 4. Nach GroupUID + Identity zählen
  const groupByUID = countIdentitiesPerGroup(allMessages);
  console.log("GroupByUID:", groupByUID);

  for (const g of groups) {
    const m = groupByUID[g.groupUid?.toString() || ""];
    if (!m) continue;

    for (const identityId in m) {
      const groupMember: ModelsGroupMember = {
        messageCount: m[identityId],
        identity: await loadIdentityByUserID(identityId),
      };
      g.groupMember?.push(groupMember);
    }
  }

  // 6. Total MessageCount berechnen
  for (const group of groups) {
    group.messageCount =
      group.groupMember?.reduce((sum, m) => sum + (m.messageCount || 0), 0) ||
      0;
  }

  return groups;
}

function countIdentitiesPerGroup(
  rows: GroupMessageRow[],
): Record<string, Record<string, number>> {
  const result: Record<string, Record<string, number>> = {};

  for (const row of rows) {
    if (!result[row.groupUID]) {
      result[row.groupUID] = {};
    }
    result[row.groupUID][row.identity] =
      (result[row.groupUID][row.identity] || 0) + 1;
  }

  return result;
}

async function loadGroupsFromCSV(fileContent: Blob): Promise<ModelsGroup[]> {
  const text = await fileContent.text();
  const records = Papa.parse(text, { header: false, skipEmptyLines: true })
    .data as string[][];

  if (records.length <= 1) return [];

  const groups: ModelsGroup[] = [];
  for (let i = 1; i < records.length; i++) {
    // Skip header
    const r = records[i];
    if (r.length < 11) continue;

    const archived = r[6] === "true" || r[6] === "1";

    groups.push({
      id: r[0],
      creator: r[1],
      groupName: r[2],
      archived,
      groupUid: r[9],
      messageCount: 0,
      groupMember: [],
    });
  }

  return groups;
}

async function loadGroupMessagesFromCSV(
  fileContent: Blob,
  filename: string,
): Promise<GroupMessageRow[]> {
  const text = await fileContent.text();
  const records = Papa.parse(text, { header: false, skipEmptyLines: true })
    .data as string[][];

  if (records.length <= 1) return [];

  const base = filename.replace(/^group_message_/, "").replace(".csv", "");
  const rows: GroupMessageRow[] = [];

  for (let i = 1; i < records.length; i++) {
    // Skip header
    const r = records[i];
    if (r.length < 12 || r[10] !== "TEXT") continue;

    const createdAtMillis = parseInt(r[7], 10);
    if (isNaN(createdAtMillis)) continue;

    const identity = r[2] || "";

    rows.push({
      groupUID: base,
      identity,
      timestamp: new Date(createdAtMillis),
    });
  }

  return rows;
}
